In [25]:
pip install tensorflow

     |████████████████████████████████| 320.4 MB 22 kB/s s eta 0:00:01�█████████▍                 | 144.4 MB 93.7 MB/s eta 0:00:02/s eta 0:00:02     |████████████████▊               | 168.0 MB 43.4 MB/s eta 0:00:04��█               | 170.8 MB 43.4 MB/s eta 0:00:04| 184.7 MB 43.4 MB/s eta 0:00:04��████▌           | 204.7 MB 39.7 MB/s eta 0:00:03:00:03 eta 0:00:03��█████████████████       | 250.0 MB 30.5 MB/s eta 0:00:03�██▋      | 256.1 MB 30.5 MB/s eta 0:00:03MB/s eta 0:00:03��█████████████▊    | 277.7 MB 87.2 MB/s eta 0:00:01��████████████████████████▎   | 283.3 MB 87.2 MB/s eta 0:00:01███████████████████   | 289.5 MB 87.2 MB/s eta 0:00:01MB 67.5 MB/s eta 0:00:01     |██████████████████████████████▌ | 305.9 MB 67.5 MB/s eta 0:00:01     |██████████████████████████████▉ | 308.2 MB 67.5 MB/s eta 0:00:01��███████████▊| 317.2 MB 67.5 MB/s eta 0:00:01
     |████████████████████████████████| 65 kB 5.6 MB/s  eta 0:00:01
     |████████████████████████████████| 57 kB 6.3 MB/s  eta 0:00:01
     

In [26]:
pip install keras

Note: you may need to restart the kernel to use updated packages.


Load the data.

In [29]:
import pandas as pd
url="https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_4d06cc69d10542e48d61fc129673b0ce.csv"
df=pd.read_csv(url,header=None)
df.head()

0   1   2   3  4   5   6   7   8   9   ...  47  48  49  50  51  52  53  54  \
0   1   0   3   0  ?   0   2   2   2   1  ...   2   2   2   2   2   1   1   1   
1   1   0   3   3  1   0   3   1   3   1  ...   2   2   2   2   2   2   2   1   
2   1   0   3   3  2   0   3   3   3   1  ...   2   2   2   2   2   2   2   2   
3   1   0   2   3  2   1   3   3   3   1  ...   2   2   2   2   2   2   2   2   
4   1   0   3   2  1   1   3   3   3   2  ...   2   2   2   2   2   2   2   1   

   55  56  
0   2   2  
1   2   2  
2   1   2  
3   2   2  
4   2   2  

[5 rows x 57 columns]

In [30]:
df.shape

(32, 57)

A high-dimension small dataset.

The associate task is classification. Separate the column that contains the labels.

In [31]:
y=pd.DataFrame(df.iloc[:,0])
y.head()

0
0  1
1  1
2  1
3  1
4  1

Check for null value in the label column.

In [32]:
y.isna().any()

0    False
dtype: bool

Separate the predictor variables.

In [33]:
X=df.iloc[:,1:57]
X.head()

1   2   3  4   5   6   7   8   9   10  ...  47  48  49  50  51  52  53  54  \
0   0   3   0  ?   0   2   2   2   1   1  ...   2   2   2   2   2   1   1   1   
1   0   3   3  1   0   3   1   3   1   1  ...   2   2   2   2   2   2   2   1   
2   0   3   3  2   0   3   3   3   1   1  ...   2   2   2   2   2   2   2   2   
3   0   2   3  2   1   3   3   3   1   2  ...   2   2   2   2   2   2   2   2   
4   0   3   2  1   1   3   3   3   2   2  ...   2   2   2   2   2   2   2   1   

   55  56  
0   2   2  
1   2   2  
2   1   2  
3   2   2  
4   2   2  

[5 rows x 56 columns]

Outputting the first five rows of the predictor variables shows that there are missing values marked with "?".

In [34]:
X.dtypes

1      int64
2      int64
3      int64
4     object
5      int64
6      int64
7      int64
8      int64
9      int64
10     int64
11     int64
12     int64
13     int64
14     int64
15     int64
16     int64
17     int64
18     int64
19     int64
20     int64
21     int64
22     int64
23     int64
24     int64
25     int64
26     int64
27     int64
28     int64
29     int64
30     int64
31     int64
32     int64
33     int64
34     int64
35     int64
36     int64
37     int64
38    object
39     int64
40     int64
41     int64
42     int64
43     int64
44     int64
45     int64
46     int64
47     int64
48     int64
49     int64
50     int64
51     int64
52     int64
53     int64
54     int64
55     int64
56     int64
dtype: object

In [35]:
dtype=pd.DataFrame(X.dtypes).rename(columns={0:"Type"})
dtype[dtype["Type"]=="object"]

Type
4   object
38  object

Every column is of datatype "int64". Column 4 and column 38 are of data type "object". Those are the two columns with missing values marked by "?".

In [36]:
import numpy as np
np.asarray(X.mode().loc[0])

array([0.0, 2.0, 3.0, '1', 0.0, 2.0, 2.0, 3.0, 1.0, 1.0, 1.0, 1.0, 2.0,
       2.0, 2, 1.0, 2.0, 2.0, 0.0, 2.0, 2.0, 2.0, 2.0, 1.0, 2.0, 2.0, 2.0,
       2.0, 2.0, 1.0, 3.0, 3.0, 3.0, 1.0, 1.0, 2.0, 1, '2', 2.0, 2.0, 1,
       2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0,
       2.0, 2.0], dtype=object)

Get the mode of column 4. Indexing in pandas starts at 0. So, the column index for column 4 in X is 3.

In [37]:
mode4=X.iloc[:,3].mode()
mode4

0    1
dtype: object

Let's see which rows in column 4 have "?".

In [38]:
Q_mark=pd.DataFrame(X.iloc[:,3]=="?").rename(columns={4:"Q_mark"})
Q_mark[Q_mark["Q_mark"]==True]

Q_mark
0     True
14    True
18    True
20    True

Rows 0, 14, 18, 20 in column 4 have "?".

Replace those "?" with the mode.

In [39]:
X_4=pd.DataFrame(X.iloc[:,3]).replace('?',1)
X_4.head()

4
0  1
1  1
2  2
3  2
4  1

Verify that the "?"'s have been replaced.

In [40]:
Q_mark=pd.DataFrame(X_4=="?").rename(columns={4:"Q_mark"})
Q_mark[Q_mark["Q_mark"]==True]

Empty DataFrame
Columns: [Q_mark]
Index: []

All the "?"'s in column 4 have been replaced with the mode.

Get the mode of the other column, column 38, that has "?"'s.

In [41]:
mode38=X.iloc[:,37].mode()
mode38

0    2
dtype: object

In [42]:
Q_mark=pd.DataFrame(X.iloc[:,37]=="?").rename(columns={38:"Q_mark"})
Q_mark[Q_mark["Q_mark"]==True]

Q_mark
25    True

Only row 25 in column 38 has "?".

In [43]:
X_38=pd.DataFrame(X.iloc[:,37]).replace('?',1)
X_38.head()

38
0  2
1  1
2  2
3  2
4  1

In [44]:
Q_mark=pd.DataFrame(X_38=="?").rename(columns={38:"Q_mark"})
Q_mark[Q_mark["Q_mark"]==True]

Empty DataFrame
Columns: [Q_mark]
Index: []

All the "?"'s in column 38 have been replaced with the mode.

Drop the columns with "?" from the original set X and concatenate the remaining columns with the cleaned results from above.

In [45]:
X1=X.iloc[:,0:3]
X2=X.iloc[:,4:37]
X3=X.iloc[:,38:57]
X=pd.concat([X1,X_4,X2,X_38,X3],axis=1)
X.head()

1   2   3  4   5   6   7   8   9   10  ...  47  48  49  50  51  52  53  54  \
0   0   3   0  1   0   2   2   2   1   1  ...   2   2   2   2   2   1   1   1   
1   0   3   3  1   0   3   1   3   1   1  ...   2   2   2   2   2   2   2   1   
2   0   3   3  2   0   3   3   3   1   1  ...   2   2   2   2   2   2   2   2   
3   0   2   3  2   1   3   3   3   1   2  ...   2   2   2   2   2   2   2   2   
4   0   3   2  1   1   3   3   3   2   2  ...   2   2   2   2   2   2   2   1   

   55  56  
0   2   2  
1   2   2  
2   1   2  
3   2   2  
4   2   2  

[5 rows x 56 columns]

Feature selection.

In [46]:
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import mutual_info_classif

# feature extraction. We are going to output the selection scores for all features and select the features with the highest scores.
test = SelectKBest(score_func=mutual_info_classif, k="all")
fit = test.fit(X, np.asarray(y).reshape(y.shape[0],))
features = fit.transform(X)
# summarize scores
set_printoptions(precision=3)

scores=fit.scores_
scores=pd.DataFrame(scores)
scores=scores.rename(columns={0:"scores"})
scores=scores.sort_values(by=["scores"],ascending=False)
scores

scores
19  0.314718
5   0.288220
26  0.253504
18  0.228080
36  0.214991
7   0.188492
22  0.153281
13  0.148331
39  0.135199
38  0.118373
32  0.116080
52  0.115356
41  0.110447
45  0.100201
20  0.079036
6   0.076960
53  0.071924
24  0.069715
12  0.052649
14  0.050886
51  0.050772
23  0.045560
42  0.040723
1   0.039896
44  0.019468
27  0.014751
34  0.011439
48  0.011254
9   0.007791
49  0.005631
55  0.005613
15  0.004014
50  0.000000
43  0.000000
46  0.000000
47  0.000000
40  0.000000
54  0.000000
0   0.000000
37  0.000000
16  0.000000
2   0.000000
3   0.000000
4   0.000000
8   0.000000
10  0.000000
11  0.000000
17  0.000000
35  0.000000
21  0.000000
25  0.000000
29  0.000000
30  0.000000
31  0.000000
33  0.000000
28  0.000000

Keep only features with non-zero scores.

In [47]:
scores=scores[scores["scores"]>0]
scores

scores
19  0.314718
5   0.288220
26  0.253504
18  0.228080
36  0.214991
7   0.188492
22  0.153281
13  0.148331
39  0.135199
38  0.118373
32  0.116080
52  0.115356
41  0.110447
45  0.100201
20  0.079036
6   0.076960
53  0.071924
24  0.069715
12  0.052649
14  0.050886
51  0.050772
23  0.045560
42  0.040723
1   0.039896
44  0.019468
27  0.014751
34  0.011439
48  0.011254
9   0.007791
49  0.005631
55  0.005613
15  0.004014

Select the features with non-zero scores.

In [48]:
X=X.iloc[:,scores.index]
X.head()

20  6   27  19  37  8   23  14  40  39  ...  43  2   45  28  35  49  10  \
0   2   2   2   0   1   2   2   2   2   2  ...   2   3   2   3   2   2   1   
1   0   3   3   0   2   3   2   3   2   2  ...   2   3   1   2   2   2   1   
2   0   3   3   0   1   3   2   3   2   2  ...   2   3   2   2   2   2   1   
3   0   3   2   0   1   3   2   3   2   2  ...   2   2   2   3   1   2   2   
4   0   3   3   0   2   3   2   2   2   2  ...   2   3   2   2   1   2   2   

   50  56  16  
0   2   2   1  
1   2   2   1  
2   2   2   1  
3   2   2   1  
4   2   2   2  

[5 rows x 32 columns]

Check if the class distributions are even.

In [49]:
y=y.rename(columns={0:"Class"})
classes=pd.concat([X,y],axis=1)
classes.groupby("Class").count().T

Class  1   2   3
20     9  13  10
6      9  13  10
27     9  13  10
19     9  13  10
37     9  13  10
8      9  13  10
23     9  13  10
14     9  13  10
40     9  13  10
39     9  13  10
33     9  13  10
53     9  13  10
42     9  13  10
46     9  13  10
21     9  13  10
7      9  13  10
54     9  13  10
25     9  13  10
13     9  13  10
15     9  13  10
52     9  13  10
24     9  13  10
43     9  13  10
2      9  13  10
45     9  13  10
28     9  13  10
35     9  13  10
49     9  13  10
10     9  13  10
50     9  13  10
56     9  13  10
16     9  13  10

The distributions of the three classes are fairly even. So, I will not undersample, oversample or use any synthetic data.

One-hot encode the target variable to prepare it for use in a neural network.

In [50]:
from sklearn.preprocessing import OneHotEncoder
enc=OneHotEncoder(handle_unknown="ignore")
y=pd.DataFrame(enc.fit_transform(y).toarray())
y

0    1    2
0   1.0  0.0  0.0
1   1.0  0.0  0.0
2   1.0  0.0  0.0
3   1.0  0.0  0.0
4   1.0  0.0  0.0
5   1.0  0.0  0.0
6   1.0  0.0  0.0
7   1.0  0.0  0.0
8   1.0  0.0  0.0
9   0.0  1.0  0.0
10  0.0  1.0  0.0
11  0.0  1.0  0.0
12  0.0  1.0  0.0
13  0.0  1.0  0.0
14  0.0  1.0  0.0
15  0.0  1.0  0.0
16  0.0  1.0  0.0
17  0.0  1.0  0.0
18  0.0  1.0  0.0
19  0.0  1.0  0.0
20  0.0  1.0  0.0
21  0.0  1.0  0.0
22  0.0  0.0  1.0
23  0.0  0.0  1.0
24  0.0  0.0  1.0
25  0.0  0.0  1.0
26  0.0  0.0  1.0
27  0.0  0.0  1.0
28  0.0  0.0  1.0
29  0.0  0.0  1.0
30  0.0  0.0  1.0
31  0.0  0.0  1.0

Split the data into train and test subsets.

In [51]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test=train_test_split(X, y, test_size=0.2,random_state=42)
print(X_train.shape)
print(X_test.shape)

(25, 32)
(7, 32)


Import the necessary libraries.

In [52]:
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adam
from keras.layers import Dropout

Configure the neural network.

In [53]:
length=X_train.shape[1]
num_classes=y_test.shape[1]

def classification_model():
    # create model
    model = Sequential()
    model.add(Dense(length, activation='relu', input_shape=(length,)))
    model.add(Dropout(0.1))
    model.add(Dense(150, activation='relu'))
    model.add(Dense(150, activation='relu'))
    model.add(Dense(num_classes, activation='softmax'))
    opt=Adam(lr=0.001)
    model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
    return model

Train the neural network and test it on test data.

In [62]:
model=classification_model()
model.fit(X_train,y_train,validation_data=(X_test, y_test),epochs=30, verbose=2)
model.evaluate(X_test,y_test,verbose=0)

Epoch 1/30
1/1 - 0s - loss: 1.1267 - accuracy: 0.4400 - val_loss: 1.0620 - val_accuracy: 0.4286
Epoch 2/30
1/1 - 0s - loss: 1.0045 - accuracy: 0.5600 - val_loss: 1.0294 - val_accuracy: 0.5714
Epoch 3/30
1/1 - 0s - loss: 1.0400 - accuracy: 0.4400 - val_loss: 1.0229 - val_accuracy: 0.5714
Epoch 4/30
1/1 - 0s - loss: 0.9487 - accuracy: 0.6800 - val_loss: 1.0250 - val_accuracy: 0.5714
Epoch 5/30
1/1 - 0s - loss: 1.0117 - accuracy: 0.5600 - val_loss: 1.0314 - val_accuracy: 0.4286
Epoch 6/30
1/1 - 0s - loss: 0.9373 - accuracy: 0.6000 - val_loss: 1.0387 - val_accuracy: 0.4286
Epoch 7/30
1/1 - 0s - loss: 0.9198 - accuracy: 0.7200 - val_loss: 1.0368 - val_accuracy: 0.4286
Epoch 8/30
1/1 - 0s - loss: 0.8861 - accuracy: 0.6000 - val_loss: 1.0276 - val_accuracy: 0.4286
Epoch 9/30
1/1 - 0s - loss: 0.8552 - accuracy: 0.6400 - val_loss: 1.0060 - val_accuracy: 0.4286
Epoch 10/30
1/1 - 0s - loss: 0.7864 - accuracy: 0.7600 - val_loss: 0.9789 - val_accuracy: 0.4286
Epoch 11/30
1/1 - 0s - loss: 0.7946 - a

[0.7565022706985474, 0.8571428656578064]

Accuracy on test data is 85.7%.

Let's see if standardizing the predictor variables will improve the results.

In [63]:
from sklearn import preprocessing
scaler=preprocessing.StandardScaler()
X_train=scaler.fit_transform(X_train)
X_train=pd.DataFrame(X_train)
X_train.head()

0         1         2         3         4         5         6   \
0 -0.041739 -1.479900 -0.750000  1.940323  0.421350 -0.685994  0.436436   
1  1.001741 -1.479900  1.333333 -0.547270 -1.083473 -0.685994  0.436436   
2  1.001741 -0.204124 -0.750000 -0.547270 -1.083473 -0.685994  0.436436   
3 -1.085219  1.071652  1.333333 -0.547270  0.421350  0.743161  0.436436   
4  1.001741 -1.479900 -0.750000 -0.547270  0.421350  0.743161  0.436436   

         7         8         9   ...        22        23        24        25  \
0 -1.749431 -2.155264  0.075810  ... -0.204124 -0.646162  0.204124 -2.173221   
1 -0.338600  0.089803  0.075810  ... -0.204124 -0.646162  0.204124 -0.232845   
2 -0.338600  0.089803  0.075810  ... -0.204124  1.148733  0.204124  1.707531   
3 -0.338600  0.089803  0.075810  ... -0.204124  1.148733  0.204124 -0.232845   
4  1.072232  0.089803 -1.819435  ... -0.204124  1.148733  0.204124 -0.232845   

         26        27        28        29       30        31  
0 -0.933139  0.116248  0.905753 -3.022439  0.62361  1.356801  
1  0.524891  0.116248  0.905753 -0.116248  0.62361 -0.527645  
2  0.524891  0.116248 -0.836080 -0.116248  0.62361 -0.527645  
3 -0.933139  0.116248  0.905753 -0.116248  0.62361  1.356801  
4  1.982920  0.116248 -0.836080 -0.116248  0.62361 -0.527645  

[5 rows x 32 columns]

In [64]:
X_test=scaler.fit_transform(X_test)
X_test=pd.DataFrame(X_test)
X_test.head()

0         1         2         3         4         5         6   \
0  0.811107 -0.632456 -0.408248 -0.632456  1.961161 -1.371989 -0.866025   
1 -0.324443 -0.632456 -0.408248 -0.632456 -0.784465 -0.171499  1.154701   
2  0.811107  1.581139 -0.408248  1.581139  0.588348 -1.371989 -0.866025   
3 -1.459993 -0.632456 -0.408248 -0.632456  0.588348  1.028992  1.154701   
4 -1.459993  1.581139 -0.408248 -0.632456 -0.784465  1.028992  1.154701   

         7         8    9   ...   22        23        24        25        26  \
0 -1.322876  2.449490  0.0  ...  0.0 -0.866025  2.449490 -1.341641 -0.784465   
1  1.322876 -0.408248  0.0  ...  0.0  1.154701 -0.408248  0.223607  1.961161   
2  0.000000 -0.408248  0.0  ...  0.0  1.154701 -0.408248  0.223607 -0.784465   
3 -1.322876 -0.408248  0.0  ...  0.0 -0.866025 -0.408248  0.223607 -0.784465   
4  1.322876 -0.408248  0.0  ...  0.0  1.154701 -0.408248 -1.341641  0.588348   

         27        28        29        30        31  
0  2.449490 -0.408248  0.408248 -1.581139  1.154701  
1 -0.408248 -0.408248  0.408248  0.632456  1.154701  
2 -0.408248 -0.408248  0.408248 -1.581139  1.154701  
3 -0.408248 -0.408248  0.408248  0.632456 -0.866025  
4 -0.408248 -0.408248  0.408248  0.632456 -0.866025  

[5 rows x 32 columns]

In [65]:
def classification_model():
    # create model
    model = Sequential()
    model.add(Dense(length, activation='relu', input_shape=(length,)))
    model.add(Dropout(0.1))
    model.add(Dense(80, activation='relu'))
    model.add(Dense(80, activation='relu'))
    model.add(Dense(num_classes, activation='softmax'))
    opt=Adam(lr=0.001)
    model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
    return model

In [66]:
model=classification_model()
model.fit(X_train,y_train,validation_data=(X_test, y_test),epochs=30, verbose=2)
model.evaluate(X_test,y_test,verbose=0)

Epoch 1/30
1/1 - 0s - loss: 1.2244 - accuracy: 0.4000 - val_loss: 1.1409 - val_accuracy: 0.4286
Epoch 2/30
1/1 - 0s - loss: 1.1686 - accuracy: 0.4000 - val_loss: 1.1177 - val_accuracy: 0.4286
Epoch 3/30
1/1 - 0s - loss: 1.1040 - accuracy: 0.4400 - val_loss: 1.0973 - val_accuracy: 0.5714
Epoch 4/30
1/1 - 0s - loss: 1.0422 - accuracy: 0.5600 - val_loss: 1.0797 - val_accuracy: 0.5714
Epoch 5/30
1/1 - 0s - loss: 0.9996 - accuracy: 0.5600 - val_loss: 1.0641 - val_accuracy: 0.4286
Epoch 6/30
1/1 - 0s - loss: 0.9741 - accuracy: 0.5600 - val_loss: 1.0497 - val_accuracy: 0.4286
Epoch 7/30
1/1 - 0s - loss: 0.9326 - accuracy: 0.6400 - val_loss: 1.0366 - val_accuracy: 0.4286
Epoch 8/30
1/1 - 0s - loss: 0.9068 - accuracy: 0.6400 - val_loss: 1.0245 - val_accuracy: 0.4286
Epoch 9/30
1/1 - 0s - loss: 0.8895 - accuracy: 0.7200 - val_loss: 1.0121 - val_accuracy: 0.4286
Epoch 10/30
1/1 - 0s - loss: 0.8441 - accuracy: 0.7200 - val_loss: 1.0006 - val_accuracy: 0.4286
Epoch 11/30
1/1 - 0s - loss: 0.8265 - a

[0.761165201663971, 0.8571428656578064]

Standardization improved the results on train data, but not on test data. Accuracy on test data is still 85.7%. 92% accuracy on train data and 85.7% accuracy on test data is quite good already. The highest accuracy from past research on this dataset that I know of is 77%. 